### PR TITLE
LB: introduce randomization in locality LB scheduler initialization

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: LB: introduce randomization in locality LB scheduler initialization
Additional Description:
Following up on the randomized picking introduced in #31592, this PR plumbs a seed into the locality-LB.
Prior to this PR whenever a new EDS assignment arrived, the order of the picked locality was always reset.
For example, if there are 2 localities, locality_A with weight 99 and locality_B with weight 1, an update of the EDS message (that doesn't update the weights) may result in starting from the beginning of the "pick list".
This may cause a skew in the traffic distribution, especially in a fleet of Envoys, and when the Envoys receive periodic assignment updates and a very low QPS.

The change ensures that the starting point of which locality to choose from is randomized and picked from some random starting point.

Risk Level: Medium - a bug fix that may impact current traffic flows.
Testing: Added tests, and updated old ones.
Docs Changes: None.
Release Notes: Added.
Platform Specific Features: N/A.
Runtime guard: Added `envoy.reloadable_features.edf_lb_locality_scheduler_init_fix`

### CODE REVIEW

**Comment by adisuissa** [comment]
Assigning load-balancing relevant maintainers.
This PR is mostly plumbing (unfortunately needed in many places of the code) and tests.
/assign @wbpcode @nezdolik @htuch

File: source/common/upstream/upstream_impl.cc
Comment (tonya11en) [inline source/common/upstream/upstream_impl.cc:638]: Do we still want to create a scheduler even if all weights are equal?

File: test/extensions/load_balancing_policies/subset/subset_test.cc
Comment (tonya11en) [inline test/extensions/load_balancing_policies/subset/subset_test.cc:1709]: Would it make more sense to just use a real random generator in these tests, as opposed to one that just returns 0?

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:638]: AFAICT this was the behavior prior to this PR, and I haven't modified that.

This comment (and code) is looking at the case where all weights are *zero*. The end-result of this method (both prior to this change and after this change) is that if all weights are zero, the `locality_scheduler` will be null.

File: test/extensions/load_balancing_policies/subset/subset_test.cc
Comment (adisuissa) [reply test/extensions/load_balancing_policies/subset/subset_test.cc:1709]: The issue is that the same RNG is used in this test (for example `EXPECT_CALL(random_, random()).WillOnce(Return(0)).WillOnce(Return(9999)).WillOnce(Return(2));` a few lines above), and I think it needs to return the specified values in some cases to ensure that the test passes.

File: source/common/upstream/upstream_impl.cc
Comment (nezdolik) [inline source/common/upstream/upstream_impl.cc:643]: do we need to ensure that `seed` > 0?

**Review by nezdolik — COMMENTED** [review]
if `edf_lb_locality_scheduler_init_fix` flag is now on by default, we need to test both code branches in the tests

File: source/common/upstream/cluster_manager_impl.h
Comment (adisuissa) [reply source/common/upstream/cluster_manager_impl.h:604]: I'm not sure if this is needed, as the seed is a POD (similar to priority above).
Even if it is modified, it won't change the caller's value.

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:628]: Yes you are correct. I opted to make the code more readable and maintainable (easy to remove the old-code path) rather than refactor the code.

Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:628]: Hmmm... good question.
I view this as a bug-fix, and so I updated the release-notes. If there's a doc you think I should update, please point me to it.

File: source/common/upstream/health_discovery_service.cc
Comment (adisuissa) [reply source/common/upstream/health_discovery_service.cc:529]: I think it is quite challenging to sync the time across a fleet, especially if we are thinking in the context of EDS updates that are being sent to many Envoys (and each triggers the update at a slightly different time).
BTW I'm also basing this code on the comment that already exists in:
https://github.com/envoyproxy/envoy/blob/b34d122fba424642f917ca64dcd5362670986373/source/common/upstream/load_balancer_impl.h#L540-L545

File: test/common/upstream/upstream_impl_test.cc
Comment (adisuissa) [reply test/common/upstream/upstream_impl_test.cc]: Good idea, done!

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:644]: `seed` is defined as `uint64_t` in the input of the current function, and as a `uint32_t` in the input of `createWithPick()`, so both are unsigned and should be >0. The implicit casting seems fine to me, as it just specifies the number of initial-picks which if it is more than 10^6 is probably bad anyway.

File: test/common/upstream/upstream_impl_test.cc
Comment (adisuissa) [reply test/common/upstream/upstream_impl_test.cc:5607]: You are correct. I've updated the test name and its description.
The randomization correctness is validated in the edf_scheduler tests.

I've also added a test that runs the old-code-path (mainly to increase coverage).

**Comment by adisuissa** [comment]
> if `edf_lb_locality_scheduler_init_fix` flag is now on by default, we need to test both code branches in the tests

In some cases we can do a parameterized test that works with both. I think the question is what is the added value of this, and in this case I'm not sure the benefit outweighs the cost.
I'm not opposing doing this, just looking for the reasons to increase test time.

**Comment by adisuissa** [comment]
Thanks for the detailed review and comments @nezdolik!

File: source/common/upstream/cluster_manager_impl.h
Comment (nezdolik) [reply source/common/upstream/cluster_manager_impl.h:604]: oh right...

**Comment by nezdolik** [comment]
> > if `edf_lb_locality_scheduler_init_fix` flag is now on by default, we need to test both code branches in the tests
> 
> In some cases we can do a parameterized test that works with both. I think the question is what is the added value of this, and in this case I'm not sure the benefit outweighs the cost. I'm not opposing doing this, just looking for the reasons to increase test time.

Would not executing code from both if branches affect coverage?

File: source/common/upstream/upstream_impl.cc
Comment (nezdolik) [reply source/common/upstream/upstream_impl.cc:628]: Sounds good to me, i don't have any particular doc in mind for this change.

**Review by nezdolik — DISMISSED** [review]

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:644]: This was validated in [this test](https://github.com/envoyproxy/envoy/blob/924f778c9ad8c5a4d8d1c6af2a54012b238a9b14/test/common/upstream/edf_scheduler_test.cc#L180).

**Comment by adisuissa** [comment]
> > > if `edf_lb_locality_scheduler_init_fix` flag is now on by default, we need to test both code branches in the tests
> > 
> > 
> > In some cases we can do a parameterized test that works with both. I think the question is what is the added value of this, and in this case I'm not sure the benefit outweighs the cost. I'm not opposing doing this, just looking for the reasons to increase test time.
> 
> Would not executing code from both if branches affect coverage?

Yes. I added a specific test that sets the runtime-flag to false, to validate that it works as expected, and that will cover the old-path code.
If you think there are other tests that may benefit from that, let me know which.

**Comment by nezdolik** [comment]
> > > > if `edf_lb_locality_scheduler_init_fix` flag is now on by default, we need to test both code branches in the tests
> > > 
> > > 
> > > In some cases we can do a parameterized test that works with both. I think the question is what is the added value of this, and in this case I'm not sure the benefit outweighs the cost. I'm not opposing doing this, just looking for the reasons to increase test time.
> > 
> > 
> > Would not executing code from both if branches affect coverage?
> 
> Yes. I added a specific test that sets the runtime-flag to false, to validate that it works as expected, and that will cover the old-path code. If you think there are other tests that may benefit from that, let me know which.

np, sorry I missed that test, it was lots of affected test code :)

File: source/common/upstream/upstream_impl.cc
Comment (wbpcode) [inline source/common/upstream/upstream_impl.cc:642]: This is LGTM overall. 

I have only one question. The `createWithPicks` will use the `seed % 429496729` as the pre picks number. 
And because the seed is a random value from `random()` method, so it's possible we may do  429496729 times pre-picks when we update the hosts.

Do we acutally evalute the possible performance impact at the worst case? If the users update the host set frequently, will this bring huge performance burden?

(PS: the host scheduler will use `seed % hosts.size()` as the pre picks number, I think this should so some similar things? like we can use `seed % locality_entries.size()` or `std::max(8, seed % locality_entries.size())` here?

**Review by wbpcode — CHANGES_REQUESTED** [review]
LGTM overall. I mark it as `request changes` because there is an important quesion to be checked first.

File: source/common/upstream/upstream_impl.cc
Comment (htuch) [inline source/common/upstream/upstream_impl.cc:1779]: Just to confirm, we're not thinking of any situations in which the seed should be set by the control plane? For WRR this is fine to randomize but consistent hashes might want this. Ignore this comment if this is only applicable to WRR style and similar ilk.

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:642]: > This is LGTM overall.
> 
> I have only one question. The `createWithPicks` will use the `seed % 429496729` as the pre picks number. And because the seed is a random value from `random()` method, so it's possible we may do 429496728 times pre-picks when we update the hosts.
> 
> Do we acutally evalute the possible performance impact at the worst case? If the users update the host set frequently, will this bring huge performance burden?
> 

While conceptually the number of pre-picks could be large, the code is bounded by the number of entries/weights (N- see [here](https://github.com/envoyproxy/envoy/blob/20c7368afa9d686a109f9601ae1b9b6028b74b0a/source/common/upstream/edf_scheduler.h#L136)), and the operation has the complexity of O(N*log(N)), which is the same as it is right now (because the localities are added one after the other to the priority-queue).
A proof of the bounded number of iterations can be found at the doc linked for [this comment](https://github.com/envoyproxy/envoy/pull/31592#issue-2063014213).

> (PS: the host scheduler will use `seed % hosts.size()` as the pre picks number, I think this should be some similar things? like we can use `seed % locality_entries.size()` or `std::max(8, seed % locality_entries.size())` here?

Yes, the next PR will be to fix the wrong initialization there as well.

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:1779]: Nope, the seed is not coming from the control-plane (the config doesn't have a knob for setting this).

IIUC consistent hashes (maglev and ring-hash) are not using the EDF-scheduler, but they are thread-aware LBs that normalize the locality+host weight into a [single normalized weight](https://github.com/envoyproxy/envoy/blob/20c7368afa9d686a109f9601ae1b9b6028b74b0a/source/common/upstream/thread_aware_lb_impl.cc#L85), and populate the data-structures using the normalized-weights.
From a first pass it seems that this code might be invoked also when consistent-hashes are used (and not sure why there's a need for an EDF-based locality scheduler in this case), so let me take a deeper look and I'll get back to you with an answer.

/wait-any

File: source/common/upstream/upstream_impl.cc
Comment (wbpcode) [reply source/common/upstream/upstream_impl.cc:642]: I see. I actually not completely get it. I will read the code more detailedly. Will unblock the PR first.

**Review by wbpcode — DISMISSED** [review]

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:1779]: Ok, I've validated that this code indeed is called when a maglev/ring-hash LB policy is used, so the EDF-scheduler is created, but it is not being used.
Seems wasteful IMHO, and could probably be optimized to not be done/allocated at all. Created #32145 to track this.

File: source/common/upstream/upstream_impl.cc
Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:642]: feel free to reach out directly and I can try to give a guided explanation.

Comment (wbpcode) [reply source/common/upstream/upstream_impl.cc:642]: I got it. Sorry for the previous negligenc.

I have to say, it's awesome 👍.

File: source/common/upstream/upstream_impl.cc
Comment (htuch) [reply source/common/upstream/upstream_impl.cc:1779]: OK, that doees seem weird, good catch.

File: source/common/upstream/upstream_impl.cc
Comment (htuch) [inline source/common/upstream/upstream_impl.cc:593]: One other question that came up while reading this PR. We're threading seed around in a lot of places. Why can't it live as a constant (for the LB) for the cluster or even server context?

Comment (adisuissa) [reply source/common/upstream/upstream_impl.cc:593]: That's a good question.
While working on this PR I had the following questions on my mind:
1. What is the granularity of the seed (e.g., is it for a single cluster, or for all clusters)?
2. Should the seed be kept in memory (say a data-member of some class)? for example initialized once and then used throughout the lifetime of the process.
My observation was that using a constant seed (let's say throughout the duration of an Envoy process) will end up with a wrong randomization. Specifically for low-QPS scenarios where EDS updates do not change the weights drastically, one may end up with an incorrect traffic pattern.
I've also decided against keeping the seed in memory (data member), but rather pass it as an argument to the function, as I'm not sure if "spending" memory for the seed is necessary. Moreover, adding it to the `HostSetImpl` class and passing using it's c'tor would have probably ended up with a similar "plumbing" work.

I think the current approach allows the code to be changed later and the seed can be saved in some context and passed just as it is being passed now.

If the concern is the number of parameters for the function (which is indeed large), then I wonder whether the function needs to be refactored somehow by splitting it to multiple ones, or if the arguments should be passed in a struct (IMHO the former is the better approach, and the latter doesn't really solve the problem).

Comment (htuch) [reply source/common/upstream/upstream_impl.cc:593]: It's probably fine, just lots of plumbing. Independent of this, it might make sense to move some of these extensively plumbed params that are appearing in many places to an options struct.

**Review by htuch — DISMISSED** [review]

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**
